### PR TITLE
collect-logs fix memory usage (SC-1590)

### PR DESCRIPTION
**Pull request by a-dubs**
## Proposed Commit Message

```
collect-logs now streams certain outputs to avoid large memory usage

_stream_command_output_to_file() was added to mimic functionality of 
_write_command_output_to_file() except instead of reading the output 
of subprocess calls into memory, it streams the outputs directly to 
the target file. this new function is used when the output of a 
subprocess call does not need to be saved into a variable.

As far as usage goes, the main difference between the two functions
is that the stream function does not return result of the subprocess
call, while the write function does.

Fixes GH-3994
LP: #1980150 
```

## Additional Context
See GitHub issue #3994 for more context on the issue this PR seeks to address.

## Test Steps
See [comment below](https://github.com/canonical/cloud-init/pull/4289#issuecomment-1664636829) for testing results and instructions to replicate / test yourself.

## Checklist:

 - [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/development/contributing.html)
 - [x] I have updated or added any unit tests accordingly
 - [x] I have updated or added any documentation accordingly

## Things left to do:
- [x] update unit tests for _write_command_output_to_file() to test that stderr outputs correctly to file 
  - [x] make any changes (if necessary) to _write_command_output_to_file() so that it handles stderr output properly
- [x] update unit tests for _stream_command_output_to_file() to test that stderr outputs correctly to file
  - [x] make any changes (if necessary) to _stream_command_output_to_file() so that it handles stderr output properly

### CODE REVIEW

**Comment by holmanb** [comment]
Do you have any tests that demonstrate that this change makes cloud-init consume less memory? If so could you please attach them here?

**Review by holmanb — CHANGES_REQUESTED** [review]
Proposed changes for improving performance should include measurements.

**Comment by a-dubs** [comment]
# Memory Testing Information

## Patch to apply in logs.py:
```
diff --git a/cloudinit/cmd/devel/logs.py b/cloudinit/cmd/devel/logs.py
index ccfb0d16..382ea307 100755
--- a/cloudinit/cmd/devel/logs.py
+++ b/cloudinit/cmd/devel/logs.py
@@ -245,6 +245,24 @@ def collect_logs(tarfile, include_userdata: bool, verbosity=0):
             verbosity=verbosity,
         )
 
+        ## TEMPORARY! : FOR TESTING MEMORY USAGE
+        subprocess.call('tr -dc "A-Za-z 0-9" < /dev/urandom | fold -w100|head -n 1000000 > bigfile.txt', shell=True)
+        test_new_func = False
+        if not test_new_func:
+            result = _write_command_output_to_file(
+                cmd=["cat", "bigfile.txt"],
+                filename=os.path.join("/tmp", "bigfile_copy.txt"),
+                msg="bigfile.txt",
+                verbosity=1,
+            )
+        else:
+            _stream_command_output_to_file(
+                cmd=["cat", "bigfile.txt"],
+                filename=os.path.join("/tmp", "bigfile_copy.txt"),
+                msg="bigfile.txt",
+                verbosity=1,
+            )
+
         for log in CLOUDINIT_LOGS:
             _collect_file(log, log_dir, verbosity)
         if include_userdata:
```

## Explanation of purpose of the included memory test patch:
- Adds snippet of code to collect-logs that does the following:
  1. First, using subprocess, creates ~100MB file called bigfile.txt
  2. Then, selects one of the two functions to test (based on the `test_new_func` variable)
  3. It then uses the selected function to write the output of calling `cat bigfile.txt` to bigfile_copy.txt.
  4. Since bigfile.txt is ~100MB, this should cause the process to use a lot of memory 
  if the contents of the file are read into memory before being written to the file.

## Test Steps / Instructions
1. First, apply the provided patch to logs.py (`cloudinit/cmd/devel/logs.py`)
2. Once you've applied the given patch, to run the tests, launch an lxd container and then shell into it and run the following command:
`/usr/bin/time -v cloud-init collect-logs`
   - This will run the cloud-init collect-logs command which will run collect-logs which will use the patch of memory testing code that was added 
   - `/usr/bin/time -v` portion of the command will print out the memory usage of the process
along with other performance metrics.
3. look for the line that's labeled `Maximum resident set size (kbytes)`  - this is the memory usage in KB

## Memory Test Results & Conclusion:

### TL:DR;
- collect-logs with `_write_command_output_to_file()` function selected used **~335MB** of memory
- collect-logs with `_stream_command_output_to_file()` function selected used **~40MB** of memory
- collect-logs with no patch applied used **~40MB** of memory
- Therefore, with _stream_command_output_to_file() selected, the memory usage of collect-logs 
is the same, despite it writing 100MB of data to a file.

### Full Raw Results of Memory Tests:

#### default performance of collect-logs (no patch applied):
```
Wrote /root/cloud-init.tar.gz
        Command being timed: "cloud-init collect-logs"
        User time (seconds): 0.75
        System time (seconds): 0.14
        Percent of CPU this job got: 94%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 0:00.95
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 40656
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 6
        Minor (reclaiming a frame) page faults: 20974
        Voluntary context switches: 74
        Involuntary context switches: 223
        Swaps: 0
        File system inputs: 2592
        File system outputs: 1144
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0
```

#### _write_command_output_to_file() perfromance test results:
```
collected bigfile.txt
Wrote /root/cloud-init.tar.gz
        Command being timed: "cloud-init collect-logs"
        User time (seconds): 3.05
        System time (seconds): 1.98
        Percent of CPU this job got: 118%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 0:04.24
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 335652 <------ MEMORY USAGE (KB)
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 10
        Minor (reclaiming a frame) page faults: 119908
        Voluntary context switches: 50529
        Involuntary context switches: 1260
        Swaps: 0
        File system inputs: 2752
        File system outputs: 395712
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0
```

#### _stream_command_output_to_file() perfromance test results:
```
collected bigfile.txt
Wrote /root/cloud-init.tar.gz
        Command being timed: "cloud-init collect-logs"
        User time (seconds): 2.88
        System time (seconds): 1.66
        Percent of CPU this job got: 119%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 0:03.80
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 40652  <------ MEMORY USAGE (KB)
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 10
        Minor (reclaiming a frame) page faults: 21440
        Voluntary context switches: 49597
        Involuntary context switches: 1185
        Swaps: 0
        File system inputs: 2752
        File system outputs: 198408
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0
```

File: cloudinit/cmd/devel/logs.py [hunk 1]
```diff
@@ -143,15 +144,31 @@ def _copytree_rundir_ignore_files(curdir, files):
...
+    output = ""
     try:
-        out, _ = subp(cmd)
+        ensure_dir(os.path.dirname(filename))
+        output = subp(cmd)[0]
+        write_file(filename, output)
+    except ProcessExecutionError as e:
+        write_file(filename, str(e))
+        _debug("collecting %s failed.\n" % msg, 1, verbosity)
+        return str(e)
```
Comment (TheRealFalcon) [inline cloudinit/cmd/devel/logs.py:155]: Do we want to return the exception text here? How would it be used?

File: cloudinit/cmd/devel/logs.py [hunk 2]
```diff
@@ -143,15 +144,31 @@ def _copytree_rundir_ignore_files(curdir, files):
 
 def _write_command_output_to_file(cmd, filename, msg, verbosity):
     """Helper which runs a command and writes output or error to filename."""
+    output = ""
     try:
-        out, _ = subp(cmd)
+        ensure_dir(os.path.dirname(filename))
+        output = subp(cmd)[0]
```
Comment (TheRealFalcon) [inline cloudinit/cmd/devel/logs.py:150]: This is the only line that can raise a `ProcessExecutionError`, so we shouldn't need anything else in the try

File: cloudinit/cmd/devel/logs.py [hunk 3]
```diff
@@ -143,15 +144,31 @@ def _copytree_rundir_ignore_files(curdir, files):
 
 def _write_command_output_to_file(cmd, filename, msg, verbosity):
     """Helper which runs a command and writes output or error to filename."""
+    output = ""
```
Comment (TheRealFalcon) [inline cloudinit/cmd/devel/logs.py:147]: `output` shouldn't need a separate declaration here

File: cloudinit/cmd/devel/logs.py [hunk 4]
```diff
@@ -143,15 +144,31 @@ def _copytree_rundir_ignore_files(curdir, files):
...
+        _debug("collected %s\n" % msg, 1, verbosity)
+        return output
+
+
+def _stream_command_output_to_file(cmd, filename, msg, verbosity):
+    """Helper which runs a command and writes output or error to filename."""
+    try:
+        ensure_dir(os.path.dirname(filename))
+        with open(filename, "w") as f:
+            subprocess.call(cmd, stdout=f, stderr=f)
```
Comment (TheRealFalcon) [inline cloudinit/cmd/devel/logs.py:166]: `ProcessExecutionError` is defined by cloud-init, not subprocess. You need to ensure you're catching the proper exceptions.

File: tests/unittests/cmd/devel/test_logs.py [hunk 5]
```diff
@@ -194,6 +209,77 @@ def fake_subp(cmd):
...
+            msg="",
+            verbosity=1,
+        )
+
+        assert test_str + "\n" == return_output1
+        assert test_str + "\n" == load_file(output_file1)
+        assert expected_stderr == return_output2
+        assert expected_stderr == load_file(output_file2)
+
+    def test_stream_command_output_to_file(self, m_getuid, tmpdir):
```
Comment (TheRealFalcon) [inline tests/unittests/cmd/devel/test_logs.py:246]: If this test is going to mostly be a duplicate of the one above it, it'd be a great candidate for [parametrization](https://docs.pytest.org/en/7.3.x/how-to/parametrize.html)

**Review by TheRealFalcon — COMMENTED** [review]
Thanks for providing the evidence requested. I left some inline comments (I know it's still WIP), but the general approach here looks good to me. Let me know when it is ready for review again.

File: cloudinit/cmd/devel/logs.py (hunk 2 above)
Comment (a-dubs) [reply cloudinit/cmd/devel/logs.py:150]: To clarify, do you mean that the ensure_dir command (line 149) should go before (outside) the try block?

**Comment by a-dubs** [comment]
Tests have been parameterized. Suggestions for tweaks to source code have been taken into account. This is no longer a WIP and should be good to go.

File: cloudinit/cmd/devel/logs.py (hunk 4 above)
Comment (holmanb) [reply cloudinit/cmd/devel/logs.py:164]: @a-dubs - This comment still needs to be addressed.

**Review by holmanb — CHANGES_REQUESTED** [review]
Thanks for the updates @a-dubs. I left a nitpick for readability, not a critical change though.

James' comment about `subprocess.call()` still needs to be addressed, as well as the accidentally commented line, but otherwise everything looks good to me!

File: cloudinit/cmd/devel/logs.py [hunk 6]
```diff
@@ -196,34 +211,36 @@ def collect_logs(tarfile, include_userdata: bool, verbosity=0):
         return 1
     tarfile = os.path.abspath(tarfile)
     log_dir = datetime.utcnow().date().strftime("cloud-init-logs-%Y-%m-%d")
+    # def _write_command_output_to_file(cmd, filename, msg, verbosity):
```
[the remark this answers is not in the document]
Comment (a-dubs) [reply cloudinit/cmd/devel/logs.py:214]: Nope! Deleted that line.

File: cloudinit/cmd/devel/logs.py [hunk 7]
```diff
@@ -143,15 +144,29 @@ def _copytree_rundir_ignore_files(curdir, files):
 
 def _write_command_output_to_file(cmd, filename, msg, verbosity):
     """Helper which runs a command and writes output or error to filename."""
+    ensure_dir(os.path.dirname(filename))
     try:
-        out, _ = subp(cmd)
+        output = subp(cmd)[0]
```
[the remark this answers is not in the document]
Comment (a-dubs) [reply cloudinit/cmd/devel/logs.py:149]: Patch has been applied. Thanks!

File: cloudinit/cmd/devel/logs.py (hunk 4 above)
Comment (a-dubs) [reply cloudinit/cmd/devel/logs.py:164]: Replaced `ProcessExecutionError` with `OSError` as this is the error that subp() catches to raise its `ProcessExecutionError`. This way `_stream_command_output_to_file` will only catch the same errors that `_write_command_output_to_file`, upholding the current expected functionality.

**Review by holmanb — APPROVED** [review]
Looks good to me!

The docs test is still failing, but if you can rebase on the tip of main then you'll pick up the fix in https://github.com/canonical/cloud-init/pull/4366 that should get that test passing.

**Comment by holmanb** [comment]
Rebased on upstream/main to pick up fix for failing docs CI.

**Comment by holmanb** [comment]
> if you can rebase on the tip of main then you'll pick up the fix in https://github.com/canonical/cloud-init/pull/4366 that should get that test passing.

Nevermind, I did that and it looks like this is good now.

Merging

**Pull request merged**